File: другое/тинькофф.аисд/2/H.py
n = int(input())
lst = list(map(int, input().split()))

prefix = [0]
lbor = [0] * n
rbor = [0] * n

lstack = [(-1, -1)]
rstack = [(-1, n)]

for i in range(len(lst)):
    prefix.append(prefix[-1] + lst[i])

    while lstack[-1][0] >= lst[i]:
        lstack.pop()
    lbor[i] = lstack[-1][1] + 1
    lstack.append((lst[i], i))

for i in range(len(lst) - 1, -1, -1):
    while rstack[-1][0] >= lst[i]:
        rstack.pop()
    rbor[i] = rstack[-1][1] - 1
    rstack.append((lst[i], i))

ans = 0
for i in range(len(lst)):
    sm = prefix[rbor[i] + 1] - prefix[lbor[i]]
    ans = max(ans, sm * lst[i])

print(ans)

File: другое/тинькофф.аисд/2/H.ts
const n = 6;
const lst: number[] = [3, 1, 6, 4, 5, 2];

const prefix: number[] = [0];
const lbor: number[] = new Array(n).fill(0);
const rbor: number[] = new Array(n).fill(0);

const lstack: [number, number][] = [[-1, -1]];
const rstack: [number, number][] = [[-1, n]];

for (let i = 0; i < lst.length; i++) {
  prefix.push(prefix[prefix.length - 1] + lst[i]);

  while (lstack[lstack.length - 1][0] >= lst[i]) {
    lstack.pop();
  }
  lbor[i] = lstack[lstack.length - 1][1] + 1;
  lstack.push([lst[i], i]);
}

for (let i = lst.length - 1; i >= 0; i--) {
  while (rstack[rstack.length - 1][0] >= lst[i]) {
    rstack.pop();
  }
  rbor[i] = rstack[rstack.length - 1][1] - 1;
  rstack.push([lst[i], i]);
}
let ans = 0;
for (let i = 0; i < lst.length; i++) {
  const sm = prefix[rbor[i] + 1] - prefix[lbor[i]];
  ans = Math.max(ans, sm * lst[i]);
}
console.log(ans);
